[FLINK-11843] Bind lifespan of Dispatcher to leader session

tillrohrmann opened the pull request:
## What is the purpose of the change

This PR is based on #9828 and #9831.

This PR factors leader election and job recovery out of the `Dispatcher`. The motivation for this change is to reduce the `Dispatcher's` mutable state and to limit the lifespan of a `Dispatcher` instance to the lifespan of the leader session. That way, the `Dispatcher` no longer needs to be reused across leader sessions which makes the component much simpler and easier to maintain.

The responsibility of the leader election is now handled by the `DefaultDispatcherRunner` which starts a `DispatcherLeaderProcess` once it is granted leadership.

The `DispatcherLeaderProcess` has two implementations: `JobDispatcherLeaderProcess` and `SessionDispatcherLeaderProcess`. The former simply starts a `DispatcherService` which is a wrapper for the `Dispatcher` with a predetermined `JobGraph`. The latter first tries to recover all persisted jobs and then passes them to the constructor of the `DispatcherService`. That way, the wrapped `Dispatcher` does not have anything to do with job recovery.

The `DispatcherService` simply starts a `Dispatcher` with the provided arguments.

The `DispatcherLeaderProcess` is written in a way that one can close it at any time (revocation of the leadership). All asynchronous operations will be stopped or ignored if the component has been stopped.

## Verifying this change

- Added `DefaultDispatcherRunnerITCase`, `DefaultDispatcherRunnerTest`, `SessionDispatcherLeaderProcessTest`, `ZooKeeperDefaultDispatcherRunnerTest`
-For the original problem of FLINK-11843 `DefaultDispatcherRunnerITCase#leaderChange_withBlockingJobManagerTermination_doesNotAffectNewLeader` was added
- For the original problem of FLINK-11665 `ZooKeeperDefaultDispatcherRunnerTest#testResourceCleanupUnderLeadershipChange` was added


## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (yes)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

TisonKun commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/DispatcherLeaderProcessImplTest.java:
```diff
@@ -239,6 +246,109 @@ public void closeAsync_duringJobRecovery_preventsDispatcherServiceCreation() thr
 	}
 
+	@Test
+	public void onRemovedJobGraph_cancelsRunningJob() throws Exception {
```
IIRC method name always follows lowerCamelCase?

tillrohrmann replied:
In this test class I'm trying to follow a new naming scheme which I believe is more helpful and is described here: https://osherove.com/blog/2005/4/3/naming-standards-for-unit-tests.html. The idea is to have more descriptive names for tests.

TisonKun replied:
interesting to try it out :-)

TisonKun commented:
Thanks for opening this pull request. The description and commit history looks good. Will check out the branch and do another pass of review later.

tillrohrmann commented:
@flinkbot run travis

TisonKun commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DefaultDispatcherRunner.java:
```diff
@@ -0,0 +1,235 @@
+			previousDispatcherLeaderProcessTerminationFuture.thenRun(newDispatcherLeaderProcess::start));
+	}
+
+	private void createAndAssignNewDispatcherLeaderProcess(UUID leaderSessionID) {
```
Redundant and hard to reason due to its `void` signature. "Assign" is already implied by `=`. Could be inlined.

tillrohrmann replied:
True, this method can be removed. Will do it.

TisonKun commented:
Thanks for opening this pull request. I have one concern about the synchronization between `Dispatcher` and `JobGraphStore`.

Since close async operation and modify job graph store operation both queued in main thread, even `Dispatcher` has been revoked leadership the previous modification wasn't cancelled. Said `DispatcherRunner` lost leadership and re-granted leadership without another leader occurred(e.g., ZK connection loss), it is possibly that the previous `Dispatcher` ran modification before it was terminated.

If you also consider it is a valid problem, then we can go ahead for a solution.

TisonKun commented:
The problem above is abstractly actions in `(grant|revoke)Leadership` are asynchronous. `(grant|revoke)Leadership` are synchronized inside LeaderElectionService but we later drop the synchronization by trigger an asynchronous operation. Given that component cannot serve before it confirms leadership it might be reasonable we keep synchronization for operations in `(grant|revoke)Leadership`.

tillrohrmann commented:
> Since close async operation and modify job graph store operation both queued in main thread, even `Dispatcher` has been revoked leadership the previous modification wasn't cancelled. Said `DispatcherRunner` lost leadership and re-granted leadership without another leader occurred(e.g., ZK connection loss), it is possibly that the previous `Dispatcher` ran modification before it was terminated.

I don't fully understand which interleaving of actions you mean @TisonKun. Do you mean that the `DispatcherRunner` loses leadership while the `Dispatcher` writes a `JobGraph` to the `JobGraphStore`? I think in this case, the shut down of the `Dispatcher` will wait until the `JobGraph` has been persisted, the corresponding `JobManagerRunner` has been shut down and the `JobGraph` has been released (however not removed). Conceptually, one can think that the leadership revocation just happened after the job submission.

If you mean a different interleaving, then I would be helpful to describe it in more detail.

TisonKun commented:
Hi @tillrohrmann, sorry for replying late. What I'm thinking is that

1. Dispatcher-1 scheduled putting job graph in main thread, using JobGraphStore-1.
2. Dispatcher-1 had been revoked leadership, scheduled closing itself in main thread.
3. Dispatcher -2 has been granted leadership, start to serve, using JobGraphStore-2.
4. JobGraphStore-1 put job graph without leadership.

When I come to this pull request I think we can defer Dispatcher-2 starts until Dispatcher-1 gracefully shutdown due to we possibly use the same LeaderStore in both JobGraphStore-1 and JobGraphStore-2. But later I notice that it should be a problem when I working on FLINK-10333 and we can prevent unexpected commit without leadership by using different LeaderStore(OneShotLeaderElectionService) per leader epoch. It might be out of this pull request scope though >_<

For you information, I agree conceptually one can think that the leadership revocation just happened after the job submission. With FLINK-10333 the persistence of job graph would fail due to it is not the leader.

For `waitForTerminatingJobManager`, I wonder isn't it no waiting for "previous JobManager" because we always instance a new `Dispatcher` per leader epoch so that they don't share `jobManagerTerminationFutures`?

tillrohrmann commented:
@TisonKun the scenario you are describing cannot happen because `JobGraphWriter#putJobGraph` is called from the `Dispatcher's` main thread. This of course has other problems but it prevents the thread interleaving of your example from occurring. In the future I would like to improve this so that persisting the `JobGraph` happens asynchronously.

I think the general contract should be that if the `Dispatcher` is shut down, then no other operation spawned by the `Dispatcher` should be running. Hence, the `Dispatcher` makes sure that it properly terminates every concurrent process. This together with `DefaultDispatcherRunner#previousDispatcherLeaderProcessTerminationFuture` will make sure that the new leader will only start executing once the previous one has completely shut down if there is a single `DispatcherRunner`. 

If there are multiple `DispatcherRunner`, then you might observe the situation that the new leader has been started and that the old leader still writes a `JobGraph` to the `JobGraphStore`. I think this problem can only be solved by making the `JobGraph` storage transactional wrt the leader session.

TisonKun commented:
Thanks for your detailed explanation of `DefaultDispatcherRunner#previousDispatcherLeaderProcessTerminationFuture`. I get the idea what start-stop happens within the same DispatcherRunner and among multiple processes. Generally the idea behind this pull request makes sense to me. I'll try to arrive code details later :-)

GJL commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/StoppedDispatcherLeaderProcess.java:
```diff
@@ -0,0 +1,68 @@
+
+	@Override
+	public CompletableFuture<String> getConfirmLeaderSessionFuture() {
+		return NEVER_COMPLETED_LEADER_SESSION_FUTURE;
```
> Should this method be changed to throw `IllegalStateException`? Afaik `getConfirmLeaderSessionFuture()` shouldn't be called on `StoppedDispatcherLeaderProcess`

Ok, I might be wrong about this one.

tillrohrmann replied:
I think you are right. I will update the implementation.

TisonKun commented:
Since we fence different leader epoch with different Dispatcher instance, I think `waitForTerminatingJobManager` should be only used in closing the former instance, i.e., `onStop`. Anywhere else especially submission and recovery should not be care about it.

I'll give this PR a full pass today and tomorrow.

TisonKun commented:
> Since we fence different leader epoch with different Dispatcher instance, I think `waitForTerminatingJobManager` should be only used in closing the former instance, i.e., `onStop`. Anywhere else especially submission and recovery should not be care about it.
> 
> I'll give this PR a full pass today and tomorrow.

Nevertheless, it could be a follow up we retry on a job manager runner non-fatal failure. Currently we conceptually use one for all restart strategy that one job manager runner failure cause the whole dispatcher on fatal error.

TisonKun commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/SessionDispatcherLeaderProcess.java:
```diff
@@ -0,0 +1,283 @@
+
+	@Override
+	protected CompletableFuture<Void> onClose() {
+		// Here we need to first call super.onClose because we want to stop the DispatcherService first
```
IIRC one of the original purposes for `onClose` is releasing ourselves from calling `super.close()`

tillrohrmann replied:
True, it should be mandatory for all implementations. I will refactor it.

TisonKun commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherFactory.java:
```diff
@@ -34,7 +36,9 @@
 	 */
 	Dispatcher createDispatcher(
 		@Nonnull RpcService rpcService,
```
No `@Nonnull` I think. We don't mark `@Nonnull` in `SessionDispatcherFactory` though.

tillrohrmann replied:
Yes you are right. Will remove the annotations.

TisonKun commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:
```diff
@@ -197,19 +191,37 @@ public void onStart() throws Exception {
 			throw exception;
 		}
+
+		startRecoveredJobs();
```
`onAddedJobGraph` submits the job graph from DispatcherProcess. Could we follow that pattern in DispatcherProcess start Dispatcher and follow by submit recovered job graphs? It would exclude "recovered" concern from `Dispatcher` and keep the recovery things out side of it.

> BTW `onAddedJobGraph` as well as `JobGraphStoreListener` is another ZK specific add-on high-availability concept.

TisonKun replied:
Well I notice that `submitJob` actually have a persisting step which isn't required for a recovered job. But if so, why we do this `onAddedJobGraph`...As is the code it works due to we support update job graph in job graph store.

tillrohrmann replied:
I would like to not change this behaviour. In the long run I would rather try to get rid of the `JobGraphListener` and the `onAddedJobGraph` and `onRemovedJobGraph` calls. Ideally, the leader recovers what it can recover at the beginning and then no more `JobGraphs` from the `JobGraphStore` will be accepted.

Yes, `submitJob` works for recovered jobs because the current `JobGraphWriter` implementations would simply overwrite the old value. Instead of introducing a new RPC method for the submission of recovered jobs, I would like to change this only if needed in the future. Moreover, I hope that we can get rid of the `JobGraphListener` interface even faster. Then this would not be needed.

TisonKun replied:
I agree. I am rearranging the design document attached in FLINK-10333 for emphasizing our proposal leader store based high-availability service, which includes the removal of `JobGraphListener` since it is originally introduced for patching coordinating concurrent modification.

I am going to start a discuss thread recently to see if our community is glad to accept it. Does it sound good to you? If so, what is a good time to start this thread?(I can finish the preparation in this week) I'm afraid that the community is blocked by many ongoing threads but IMO stability(high-availability) is critical component we'd better take more priority with.

TisonKun commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:
```diff
@@ -869,78 +823,6 @@ private void jobMasterFailed(JobID jobId, Throwable cause) {
-		}
-	}
-
 	private CompletableFuture<Void> waitForTerminatingJobManager(JobID jobId, JobGraph jobGraph, FunctionWithException<JobGraph, CompletableFuture<Void>, ?> action) {
```
We don't have to wait for a previous Dispatcher terminate job manager any more. If any time we do wait for a terminating job manager to submit job, it is in a wrong state IMO.

tillrohrmann replied:
This is not strictly true atm. Imagine that someone (e.g. a test) submits the same job after the first one has completed. Since the termination can take some time, we still need to check this. Changing this now, would mean a behavioural change.

I would like to change this later as part of FLINK-11719 but not as part of this PR. The scope is already too large.

TisonKun replied:
Make sense to defer the change since it is not strongly required by this issue.

Conceptually we don't have to "wait for previous terminating job manager" because since Dispatcher spawned per leader session, if a previous terminating job manager exists, the job must be finished(successfully or unsuccessfully), which means we should reject the later submission.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/DefaultDispatcherRunnerTest.java:
```diff
@@ -0,0 +1,415 @@
+		}
+	}
+
+	private static class StartStopTestingDispatcherLeaderProcess {
```
I'll rename it to `StartStopDispatcherLeaderProcess`

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DefaultDispatcherRunner.java:
```diff
@@ -0,0 +1,235 @@
+
+	private final CompletableFuture<ApplicationStatus> shutDownFuture;
+
+	private boolean isRunning;
```
Fair enough. I'll change it. But I don't think that this is consistently applied in Flink's code base.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DefaultDispatcherRunner.java:
```diff
@@ -0,0 +1,235 @@
+		this.previousDispatcherLeaderProcessTerminationFuture = CompletableFuture.completedFuture(null);
+		this.dispatcherGatewayFuture = new CompletableFuture<>();
+
+		startDispatcherRunner(leaderElectionService);
```
You are right. I'll introduce a wrapper which is responsible for the management of the `LeaderElectionService` outside of the `DefaultDispatcherRunner`.

tillrohrmann replied:
It's called `DispatcherRunnerLeaderElectionLifecycleManager`

GJL replied:
Is it not possible to introduce a start method on the DispatcherRunner?

tillrohrmann replied:
That would be a different solution. However, I would like to avoid additional lifecycle methods as much as possible because they complicate things.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DispatcherRunner.java:
```diff
@@ -30,11 +31,12 @@
+	 * @return Future which is completed with the leader's gateway
 	 */
-	Dispatcher getDispatcher();
+	CompletableFuture<DispatcherGateway> getDispatcherGateway();
```
Valid point. I'll try to refactor the tests to get rid of this method.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DispatcherLeaderProcess.java:
```diff
@@ -0,0 +1,43 @@
+
+	CompletableFuture<DispatcherGateway> getDispatcherGateway();
+
+	CompletableFuture<String> getConfirmLeaderSessionFuture();
```
Good point. I'll rename it into `getLeaderAddressFuture`.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/AbstractDispatcherLeaderProcess.java:
```diff
@@ -0,0 +1,244 @@
+		this.shutDownFuture = new CompletableFuture<>();
+
+		this.state = State.CREATED;
+		this.dispatcherService = null;
```
True.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/AbstractDispatcherLeaderProcess.java:
```diff
@@ -0,0 +1,244 @@
+		}
+	}
+
+	protected void runIfStateIs(State expectedState, Runnable action) {
```
Good point. Will update it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/SessionDispatcherLeaderProcess.java:
```diff
@@ -0,0 +1,283 @@
+ * Process which encapsulates the job recovery logic and life cycle management of a
+ * {@link Dispatcher}.
+ */
+public class SessionDispatcherLeaderProcess extends  AbstractDispatcherLeaderProcess implements JobGraphStore.JobGraphListener {
```
Good catch. Will fix it.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/ZooKeeperDefaultDispatcherRunnerTest.java:
```diff
@@ -0,0 +1,253 @@
+			return ZooKeeperUtils.createJobGraphs(client, configuration);
+		} catch (Exception e) {
+			ExceptionUtils.rethrow(e);
+			return null;
```
Yes, let's do it as a follow up.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/ThrowingJobGraphWriter.java:
```diff
@@ -0,0 +1,42 @@
+	@Override
+	public void putJobGraph(JobGraph jobGraph) {
+		throw new UnsupportedOperationException(
+			String.format("Cannot store job graphs in the %s.", getClass().getSimpleName()));
```
True. Will update it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java:
```diff
@@ -130,6 +135,8 @@ public DispatcherRunner getDispatcherRunner() {
 	}
 
 	private CompletableFuture<Void> closeAsyncInternal() {
+		LOG.info("Closing components of {}.", getClass().getSimpleName());
```
True. I will remove this pattern from the PR.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/DefaultDispatcherRunnerTest.java:
```diff
@@ -0,0 +1,415 @@
+	public void revokeLeadership_withExistingLeader_stopsLeaderProcess() throws Exception {
+		final UUID leaderSessionId = UUID.randomUUID();
+
+		final CompletableFuture<Void> startFuture = new CompletableFuture<>();
```
True, even though it does not make a big difference imo. Will change it.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/SessionDispatcherLeaderProcessTest.java:
```diff
@@ -0,0 +1,517 @@
+
+			final CompletableFuture<Throwable> errorFuture = fatalErrorHandler.getErrorFuture();
+			final Throwable throwable = errorFuture.get();
+			Assert.assertThat(ExceptionUtils.findThrowable(throwable, expectedFailure::equals).isPresent(), Is.is(true));
```
True, it should be consistent. I will update the test case.

Why do you prefer JUnit's `assertThat` method?

GJL replied:
The only reason is to keep consistency so that all assertions come from `org.junit.Assert`, e.g., `assertTrue`, `assertEquals`, `assertThat`). Semantically there is no difference at the moment.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/SessionDispatcherLeaderProcessTest.java:
```diff
@@ -0,0 +1,517 @@
+
+			final CompletableFuture<Throwable> errorFuture = fatalErrorHandler.getErrorFuture();
+			final Throwable throwable = errorFuture.get();
+			Assert.assertThat(ExceptionUtils.findThrowable(throwable, expectedFailure::equals).isPresent(), Is.is(true));
```
Good catch. Thanks.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/SessionDispatcherLeaderProcessTest.java:
```diff
@@ -0,0 +1,517 @@
+
+			// we expect that a fatal error occurred
+			final Throwable error = fatalErrorHandler.getErrorFuture().get();
+			Assert.assertThat(ExceptionUtils.findThrowableWithMessage(error, testException.getMessage()).isPresent(), is(true));
```
Yes, will update.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/SessionDispatcherLeaderProcessTest.java:
```diff
@@ -0,0 +1,517 @@
+
+			final Collection<JobGraph> recoveredJobGraphs = recoveredJobGraphsFuture.get();
+
+			assertThat(recoveredJobGraphs, hasSize(1));
```
Good point. I will remove the unnecessary assertion.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/SessionDispatcherLeaderProcessTest.java:
```diff
@@ -0,0 +1,517 @@
+
+	@Test
+	public void closeAsync_duringJobRecovery_preventsDispatcherServiceCreation() throws Exception {
+		final OneShotLatch jobRecoveryStarted = new OneShotLatch();
```
True, I will make it consistent.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/SessionDispatcherLeaderProcessTest.java:
```diff
@@ -0,0 +1,517 @@
+		final FlinkException testException = new FlinkException("Test exception");
+		jobGraphStore = TestingJobGraphStore.newBuilder()
+			.setRecoverJobGraphFunction(
+				(JobID jobId, Map<JobID, JobGraph> jobGraphs) -> {
```
Will remove them.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/DefaultDispatcherRunnerITCase.java:
```diff
@@ -0,0 +1,234 @@
+
+			final Collection<JobID> jobIds = secondDispatcherGateway.listJobs(TIMEOUT).get();
+
+			assertThat(jobIds, hasSize(1));
```
True, will remove it.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/DefaultDispatcherRunnerTest.java:
```diff
@@ -0,0 +1,415 @@
+			shutDownFuture.complete(finalApplicationStatus);
+
+			try {
+				dispatcherShutDownFuture.get(10L, TimeUnit.MILLISECONDS);
```
The idea was to use a short timeout to allow for asynchronicity in the underlying implementation which we don't have atm. I think it can be simplified as you suggested.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/SessionDispatcherLeaderProcessTest.java:
```diff
@@ -0,0 +1,517 @@
+			assertThat(jobGraphStopFuture.isDone(), is(false));
+			assertThat(terminationFuture.isDone(), is(false));
+
+			dispatcherServiceShutdownLatch.await();
```
I think there is no good reason. Will remove it.

tillrohrmann commented:
Thanks a lot for the review @GJL and @TisonKun. I've addressed all of your comments with which I agree. On the others, I've left a comment.

As a next step, I would like to rebase on the latest master and squash the fixups if that's ok with you. The reason for this is that I expect some merge/rebase conflicts and I would like to resolve them fast.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/AbstractDispatcherLeaderProcess.java:
```diff
@@ -0,0 +1,255 @@
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+@SuppressWarnings("checkstyle:EmptyLineSeparator")
```
Will remove this line when merging.

tillrohrmann commented:
I've force pushed a rebased version with a cleaned up commit history @TisonKun and @GJL.

TisonKun commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherServices.java:
```diff
@@ -74,22 +78,24 @@ public DispatcherServices(
 			@Nonnull GatewayRetriever<ResourceManagerGateway> resourceManagerGatewayRetriever,
```
I'm afraid `org.apache.flink.runtime.dispatcher.DispatcherServices` has a name conflict with `org.apache.flink.runtime.dispatcher.runner.AbstractDispatcherLeaderProcess.DispatcherService`. Technically we can distinguish by package/namespace but it possibly brings confusion later.

TisonKun commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherServices.java:
```diff
@@ -74,22 +78,24 @@ public DispatcherServices(
 			@Nonnull GatewayRetriever<ResourceManagerGateway> resourceManagerGatewayRetriever,
```
What is `@Nonnull` annotation conventions here? Following the statement

> This means that @Nonnull annotations are usually not necessary, but can be used in certain cases to override a previous annotation, or to point non-nullability out in a context where one would expect a nullable value.

I think here one would always expect a non-null value.

tillrohrmann replied:
Yes you are right. Since this change is not part of this PR I would like to address this problem as a follow up issue.

TisonKun commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/NoOpJobGraphWriter.java:
```diff
@@ -0,0 +1,45 @@
+/**
+ * Testing implementation of {@link JobGraphWriter} which does nothing.
+ */
+public enum NoOpJobGraphWriter implements JobGraphWriter {
```
If it is a testing only class, we'd better move it to test scope.

tillrohrmann replied:
Good catch. Will move it to the `test`.

TisonKun commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/PartialDispatcherServicesWithJobGraphStore.java:
```diff
@@ -0,0 +1,86 @@
+			@Nonnull ArchivedExecutionGraphStore archivedExecutionGraphStore,
+			@Nonnull FatalErrorHandler fatalErrorHandler,
+			@Nonnull HistoryServerArchivist historyServerArchivist,
+			@Nullable String metricQueryServiceAddress,
```
Possible follow-up, metricQueryServiceAddress might be non-null [FLINK-14499](https://issues.apache.org/jira/browse/FLINK-14499). Out of this pull request but just FYI.

tillrohrmann replied:
Yes, let's address it as part of FLINK-14499.

TisonKun commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/AbstractDispatcherLeaderProcess.java:
```diff
@@ -0,0 +1,254 @@
+			JobGraphWriter jobGraphWriter);
+	}
+
+	interface DispatcherService extends AutoCloseableAsync {
```
Shall we call this one `DispatcherContainer` since it is a thin wrapper of

```
Dispatcher dispatcher;
DispatcherGateway dispatcherGateway;
```

and its methods don't like a service.

What is more important is that I'm afraid it is a bit confusing the different between this `DispatcherService` and `DispatcherServices` which are totally different thing but share a similar name under the same scope.

tillrohrmann replied:
I see your concern Tison. However, I'm not sure whether I like `DispatcherContainer` too much. My reasoning was a bit along the lines of `Executor` and `ExecutorService` where the `XService` gives you access to lifecycle methods.

tillrohrmann replied:
What about `DispatcherGatewayService`?

TisonKun replied:
Make sense. Thanks for the update.

TisonKun commented on flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DefaultDispatcherResourceManagerComponentFactory.java:
```diff
@@ -182,26 +182,25 @@ public DispatcherResourceManagerComponent create(
 
 			final HistoryServerArchivist historyServerArchivist = HistoryServerArchivist.createHistoryServerArchivist(configuration, webMonitorEndpoint);
 
-			jobManagerMetricGroup = MetricUtils.instantiateJobManagerMetricGroup(
```
We remains a `jobManagerMetricGroup` variable above. IIUC we change `jobManagerMetricGroup` to `JobManagerMetricGroupFactory` so that we can remove that variable.

tillrohrmann replied:
True, I will remove it.

GJL commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:
```diff
@@ -243,18 +255,6 @@ private void stopDispatcherServices() throws Exception {
 			exception = ExceptionUtils.firstOrSuppressed(e, exception);
```
nit: This line isn't needed anymore.

tillrohrmann replied:
True, will remove it.

GJL commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DefaultDispatcherRunner.java:
```diff
@@ -0,0 +1,197 @@
+	public void handleError(Exception exception) {
+		fatalErrorHandler.onFatalError(
+			new FlinkException(
+				String.format("Exception during leader election of %s occurred.", getClass().getSimpleName()),
```
Including `getClass().getSimpleName()` not needed but also doesn't hurt I guess.

tillrohrmann replied:
True. Will remove it.

GJL commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DispatcherRunnerLeaderElectionLifecycleManager.java:
```diff
@@ -0,0 +1,71 @@
+		try {
+			leaderElectionService.stop();
+		} catch (Exception e) {
+			exception = e;
```
You could return from here directly. Is this for extensibility?

tillrohrmann replied:
I guess it is an artifact from a previous version. I'll update it.

tillrohrmann commented:
Thanks a lot for the review @GJL. I've addressed your comments, rebased and force pushed. Merging once Travis gives green light.
